nixos/modules/security/pam.nix

```diff
@@ -460,6 +460,94 @@ let
+          options.rules = mkOption {
+            description = lib.mdDoc ''
+              PAM rules for this service and type.
+
+              Attribute keys are the name of each rule.
+            '';
+            type = types.attrsOf (types.submodule ({ config, ... }: {
+              options = {
+                enable = mkOption {
+                  type = types.bool;
+                  default = true;
+                  description = lib.mdDoc ''
+                    Whether this rule is added to the PAM service config file.
+                  '';
+                };
+                order = mkOption {
+                  type = types.int;
+                  default = 1000;
+                  description = lib.mdDoc ''
+                    Order of this rule in the service file. Rules are arranged in ascending order of this value.
+
+                    ::: {.warning}
+                    The `order` values for the built-in rules are subject to change. If you assign a constant value to this option, a system update could silently reorder your rule. You could be locked out of your system, or your system could be left wide open. If you must use this option, set it to a relative offset from another rule's `order` value:
+
+                    ```nix
+                    {
+                      security.pam.services.login.types.auth.rules.foo.order =
+                        config.security.pam.services.login.types.auth.rules.unix.order + 1;
+                    }
+                    ```
+                    :::
+                  '';
+                };
+                control = mkOption {
+                  type = types.str;
+                  description = lib.mdDoc ''
+                    Indicates the behavior of the PAM-API should the module fail to succeed in its authentication task. See `control` in {manpage}`pam.conf(5)` for details.
+                  '';
+                };
+                modulePath = mkOption {
+                  type = types.str;
+                  description = lib.mdDoc ''
+                    Either the full filename of the PAM to be used by the application (it begins with a '/'), or a relative pathname from the default module location. See `module-path` in {manpage}`pam.conf(5)` for details.
+                  '';
+                };
+                args = mkOption {
+                  type = types.listOf types.str;
+                  description = lib.mdDoc ''
+                    Tokens that can be used to modify the specific behavior of the given PAM. Such arguments will be documented for each individual module. See `module-arguments` in {manpage}`pam.conf(5)` for details.
+
+                    Escaping rules for spaces and square brackets are automatically applied.
+
+                    {option}`settings` are automatically added as {option}`args`. It's recommended to use the {option}`settings` option whenever possible so that arguments can be overridden.
+                  '';
+                };
```

Comment:
`args` is a must-have because that's the contract exposed to PAMs. The string formatting rules apply to all modules. `settings` is an optional abstraction we put on top because it makes modifying the configuration quite nice in Nix.

I had tried combining them by allowing `settings."service = systemd-user" = true` to render without the whitespace escaping. This worked for that example, but it doesn't work when there are derivation paths in the string. Splitting `args` and `settings` turned out to be more natural.

Reply:
The problem with any list-based option is that it can't be overridden really. But if that's what PAM allows then we can't avoid it :+1: